File: 22. Subarray Product Less Than K.py
# Topic: Subarray Product Less Than K

def num_subarray_product_less_than_k(nums, k):
    if k <= 1: return 0
    prod = 1
    left = 0
    ans = 0
    for right, val in enumerate(nums):
        prod *= val
        while prod >= k and left <= right:
            prod //= nums[left]
            left += 1
        ans += right - left + 1
    return ans

if __name__ == '__main__':
    nums = [10,5,2,6]
    k = 100
    print('Count:', num_subarray_product_less_than_k(nums,k))

# Time Complexity: O(n)
# Space Complexity: O(1)

File: 22. Subarray Product Less Than K.cpp
// Topic: Subarray Product Less Than K
#include <bits/stdc++.h>
using namespace std;

int numSubarrayProductLessThanK(vector<int>& nums, int k)
{
    if (k <= 1)
        return 0;
    long long prod = 1;
    int left = 0, ans = 0;
    for (int right = 0; right < nums.size(); ++right)
    {
        prod *= nums[right];
        while (prod >= k && left <= right) prod /= nums[left++];
        ans += right - left + 1;
    }
    return ans;
}

int main()
{
    vector<int> nums = {10, 5, 2, 6};
    int k = 100;
    cout << "Count: " << numSubarrayProductLessThanK(nums, k) << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)
